srml/support/procedural/src/construct_runtime/transformations.rs

```diff
@@ -0,0 +1,377 @@
+use super::parse::{ModuleDeclaration, ModulePart, RuntimeDefinition, WhereSection};
+use proc_macro::TokenStream;
+use proc_macro2::{Span, TokenStream as TokenStream2};
+use quote::quote;
+use srml_support_procedural_tools::syn_ext as ext;
+use srml_support_procedural_tools::{generate_crate_access, generate_hidden_includes};
+use syn::Ident;
+
+// try macro but returning tokenized error
+macro_rules! try_tok(( $expre : expr ) => {
+	match $expre {
+		Ok(r) => r,
+		Err (err) => {
+			return err.to_compile_error().into()
+		}
+	}
+});
+
+pub fn construct_runtime(input: TokenStream) -> TokenStream {
+	let definition = syn::parse_macro_input!(input as RuntimeDefinition);
+	let RuntimeDefinition {
+		name,
+		where_section: WhereSection {
+			block,
+			node_block,
+			unchecked_extrinsic,
+			..
+		},
+		modules: ext::Braces {
+			content: ext::Punctuated { inner: modules, .. },
+			token: modules_token,
+		},
+		..
+	} = definition;
+
+	// Assert we have system module declared
+	let system_module = match find_system_module(modules.iter()) {
+		Some(sm) => sm,
+		None => {
+			return syn::Error::new(
+				modules_token.span,
+				"`System` module declaration is missing. \
+				 Please add this line: `System: system::{Module, Call, Storage, Config, Event},`",
+			)
+			.to_compile_error()
+			.into()
+		}
+	};
+
+	let hidden_crate_name = "construct_runtime";
+	let scrate = generate_crate_access(&hidden_crate_name, "srml-support");
+	let scrate_decl = generate_hidden_includes(&hidden_crate_name, "srml-support");
+	let outer_event = try_tok!(decl_outer_event_or_origin(
+		&name,
+		modules.iter().filter(|module| module.name.to_string() != "System"),
+		&system_module,
+		&scrate,
+		DeclOuterKind::Event,
+	));
+	let outer_origin = try_tok!(decl_outer_event_or_origin(
+		&name,
+		modules.iter().filter(|module| module.name.to_string() != "System"),
```

Comment:
You don't need to call `to_string` always. Ident implements the correct `PartialEq` for that.